packages/payment-processor/src/payment/utils.ts

```diff
@@ -0,0 +1,145 @@
+  if (!signerOrProvider) {
+    signerOrProvider = getProvider();
+  }
+  if (signerOrProvider instanceof Signer) {
+    return signerOrProvider;
+  }
+  if (signerOrProvider instanceof Web3Provider) {
+    return signerOrProvider.getSigner(address);
+  }
+  throw new Error('cannot get signer');
+}
+
+/**
+ * Utility to return the payment network extension of a Request.
+ * @param request
+ */
+export function getPaymentNetworkExtension(
+  request: ClientTypes.IRequestData,
+): ExtensionTypes.IState | undefined {
+  // tslint:disable-next-line: typedef
+  return Object.values(request.extensions).find(
+    x => x.type === ExtensionTypes.TYPE.PAYMENT_NETWORK,
+  );
+}
+
+/**
+ * Utility to access the payment address and reference of a Request.
+ * @param request
+ */
+export function getRequestPaymentValues(
+  request: ClientTypes.IRequestData,
+): { paymentAddress: string; paymentReference: string } {
+  const extension = getPaymentNetworkExtension(request);
+  if (!extension) {
+    throw new Error('no payment network found');
+  }
+  const { paymentAddress, salt } = extension.values;
+  const paymentReference = PaymentReferenceCalculator.calculate(
+    request.requestId,
+    salt,
+    paymentAddress,
+  );
+  return { paymentAddress, paymentReference };
+}
+
+const { ERC20_PROXY_CONTRACT, ETH_INPUT_DATA } = PaymentTypes.PAYMENT_NETWORK_ID;
+const currenciesMap: any = {
+  [ERC20_PROXY_CONTRACT]: RequestLogicTypes.CURRENCY.ERC20,
+  [ETH_INPUT_DATA]: RequestLogicTypes.CURRENCY.ETH,
+};
+
+/**
+ * Utility to validate a request depending on the expected paymentNetwork.
+ * @param request
+ * @param paymentNetworkId
+ */
+export function validateRequest(
+  request: ClientTypes.IRequestData,
+  paymentNetworkId: PaymentTypes.PAYMENT_NETWORK_ID,
+): void {
+  const extension = request.extensions[paymentNetworkId];
+  const expectedCurrencyType = currenciesMap[paymentNetworkId];
+  if (
+    !expectedCurrencyType ||
+    request.currencyInfo.type !== expectedCurrencyType ||
+    !request.currencyInfo.network ||
+    !extension ||
+    !extension.values.salt ||
+    !extension.values.paymentAddress ||
+    (paymentNetworkId === ERC20_PROXY_CONTRACT && !request.currencyInfo.value)
+  ) {
+    throw new Error(`request cannot be processed, or is not an ${paymentNetworkId} request`);
+  }
+}
+
+/**
+ * Computes the amount to pay.
+ * If `amount` is specified, it will return it.
+ * Otherwise, it will return the amount left to pay in the request.
+ *
+ * @param request the request to pay
+ * @param amount the optional amount to pay.
+ */
+export function getAmountToPay(
+  request: ClientTypes.IRequestData,
+  amount?: BigNumberish,
+): BigNumber {
+  const amountToPay =
+    amount === undefined
+      ? bigNumberify(request.expectedAmount).sub(request.balance?.balance || 0)
+      : bigNumberify(amount);
+
+  if (amountToPay.lt(0)) {
+    throw new Error('cannot pay a negative amount');
```

Comment:
this message will be cryptic for the user.
maybe something like `the request is already paid`